### PR TITLE
join menu changes so disconnected hosts are removed.

### PR DESCRIPTION
**Pull request by theWill**
- mostly changes around removing deactivated servers from the list. 

Other changes:

- Removed initial call to UpdateItem and just made the next tick really soon after creating the server entry as the old behavior would show new servers as disconnected for the first 5 seconds (maybe data wasn't in yet?).

- made some things public like the list of servers and made serializable. Allows visibility in the inspector as to what is going on.

### CODE REVIEW

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs [hunk 1]
```diff
@@ -37,8 +38,8 @@ private void Awake()
 			// Init the MainThreadManager
 			MainThreadManager.Create();
 
-			mpMenu = this.GetComponentInParent<MultiplayerMenu>();
-			Settings = mpMenu.Settings;
+			mpMenu = FindObjectOfType<MultiplayerMenu>();
```
Comment (phalasz) [inline ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: What's the reason for this change? Ideally the Join menu should be a child of the main multiplayer menu.

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs [hunk 2]
```diff
@@ -128,7 +146,8 @@ private void LocalServerLocated(NetWorker.BroadcastEndpoints endpoint, NetWorker
 		{
 			MainThreadManager.Run(() =>
 			{
-				AddServer(endpoint.Address, endpoint.Port);
+                if (!endpoint.Address.Equals(LocalNetworkScanning.GetLocalIPAddress()))
```
Comment (phalasz) [inline ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: hmm.. wouldn't this prevent listing servers that are running on the same machine as the client?

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs [hunk 3]
```diff
@@ -81,20 +82,37 @@ private void Update()
...
+
+                        if (!server.SqpQuery.ValidResult)
+                            server.removeNextUpdate = true;
+                        else
+                            server.removeNextUpdate = false;
+
+                        server.NextUpdate = Time.time + 3.0f + UnityEngine.Random.Range(0.0f, 1.0f);
+                    }
+                }
+                foreach (ServerListItemData slid in invalidServers)
```
Comment (phalasz) [inline ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: Missing new line before this foreach.

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs [hunk 4]
```diff
@@ -161,11 +180,9 @@ private void AddServer(string address, ushort port = NetWorker.DEFAULT_PORT, boo
 			var endpoint = HostResolver.Resolve(address, Settings.SQPPort);
 
 			serverListItemData.SqpQuery = sqpClient.GetQuery(endpoint);
+            serverListItemData.NextUpdate = Time.time + .2f;
```
Comment (phalasz) [inline ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: I'd still keep a random value in there so not all updates go out at the same time.

Comment (theWill) [reply ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: I'll add it back

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs (hunk 3 above)
Comment (theWill) [reply ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: noted

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs (hunk 2 above)
Comment (theWill) [reply ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: hmmmm, perhaps so. This makes sense for my use case of having each instance start up a host automatically... I'll remove.

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs (hunk 1 above)
Comment (theWill) [reply ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: agree. I moved it in my project which is why I have this... I'll remove this.

Comment (phalasz) [reply ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: 👍

File: ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs (hunk 2 above)
Comment (phalasz) [reply ForgeUnity/Assets/BeardedManStudios/Scripts/Multiplayer Menu/JoinMenu.cs]: It is just a convenience for local testing.

**Comment by theWill** [comment]
One other thing I haven't done. I think we should move all the logic from Update() to FixedUpdate(). You can notice that the update() loop fires so quickly that it actually does multiple updates to the pings. You cool with that? @phalasz

**Comment by phalasz** [comment]
I think adding an extra second or so randomness to the timers should fix that, so it could fire as many updates, but only ping the servers when the next update time comes.

**Comment by theWill** [comment]
I am closing this. I ended up reverting to mostly use your code due to issues I had after further testing. Thanks for humoring me at the time. @phalasz
